### PR TITLE
add etag and count to key backup endpoints

### PR DESCRIPTION
**Pull request by uhoreg**
These two fields were proposed by the Riot mobile team to help clients guess if the backup has new keys since the client last uploaded keys.  They are described in [the MSC](https://github.com/matrix-org/matrix-doc/pull/1538).  ~There are some minor clarifications that need to be made to the MSC, but other than that, I think this is ready.~

~(Personally, I'm not too fond of calling the one field `hash`, since it isn't actually a hash, but I'm not bothered enough by it to try to change it at this point.)~

### CODE REVIEW

**Comment by uhoreg** [comment]
(The sytest failures are due to #5857 not being merged yet)

File: synapse/handlers/e2e_room_keys.py [hunk 1]
```diff
@@ -166,43 +182,43 @@ def upload_room_keys(self, user_id, version, room_keys):
...
+            # submitted.  Then compare them with the submitted keys.  If the
+            # key is new, insert it; if the key should be updated, then update
+            # it; otherwise, drop it.
+            existing_keys = yield self.store.get_e2e_room_keys_multi(
+                user_id, version, room_keys["rooms"]
+            )
+            to_insert = []  # batch the inserts together
+            changed = False  # if anything has changed, we need to update the hash
             for room_id, room in iteritems(room_keys["rooms"]):
-                for session_id, session in iteritems(room["sessions"]):
```
Comment (uhoreg) [inline synapse/handlers/e2e_room_keys.py]: FWIW, the diff gets hard to follow in this section, and it's much easier to just [look at the file](https://github.com/matrix-org/synapse/blob/d081883f1e2bffabbcbc94d352ca819b19fd32fd/synapse/handlers/e2e_room_keys.py#L194-L221) for this section.

File: synapse/handlers/e2e_room_keys.py [hunk 2]
```diff
@@ -166,43 +184,43 @@ def upload_room_keys(self, user_id, version, room_keys):
...
-        try:
-            current_room_key = yield self.store.get_e2e_room_key(
-                user_id, version, room_id, session_id
-            )
-        except StoreError as e:
-            if e.code == 404:
-                pass
-            else:
-                raise
+            version_hash = int(version_info["hash"])
```
[the remark this answers is not in the document]
Comment (uhoreg) [reply synapse/handlers/e2e_room_keys.py]: > What makes it so hard to change its name?

Just that the Riot mobile apps have already implemented using that name, for quite a while.  But given that it hasn't been implemented server-side, I guess changing it wouldn't break it any further.   And also that if I don't know what a better name would be. :P 

@manuroe would you mind if we changed the name of the "hash" field to something different?

File: synapse/storage/e2e_room_keys.py [hunk 3]
```diff
@@ -1,5 +1,6 @@
 # -*- coding: utf-8 -*-
 # Copyright 2017 New Vector Ltd
+# Copyright 2019 Matrix.org Foundation CIC
```
Comment (uhoreg) [inline synapse/storage/e2e_room_keys.py]:
```suggestion
# Copyright 2019 Matrix.org Foundation C.I.C.
```

File: synapse/storage/schema/delta/56/room_key_hash.sql [hunk 4]
```diff
@@ -0,0 +1,17 @@
+/* Copyright 2019 Matrix.org Foundation CIC
```
Comment (uhoreg) [inline synapse/storage/schema/delta/56/room_key_hash.sql]:
```suggestion
/* Copyright 2019 Matrix.org Foundation C.I.C.
```

File: tests/handlers/test_e2e_room_keys.py [hunk 5]
```diff
@@ -1,6 +1,7 @@
 # -*- coding: utf-8 -*-
 # Copyright 2016 OpenMarket Ltd
 # Copyright 2017 New Vector Ltd
+# Copyright 2019 Matrix.org Foundation CIC
```
Comment (uhoreg) [inline tests/handlers/test_e2e_room_keys.py]:
```suggestion
# Copyright 2019 Matrix.org Foundation C.I.C.
```

File: synapse/storage/e2e_room_keys.py [hunk 6]
```diff
@@ -226,10 +299,12 @@ def _get_e2e_room_keys_version_info_txn(txn):
                 txn,
                 table="e2e_room_keys_versions",
                 keyvalues={"user_id": user_id, "version": this_version, "deleted": 0},
-                retcols=("version", "algorithm", "auth_data"),
+                retcols=("version", "algorithm", "auth_data", "hash"),
             )
             result["auth_data"] = json.loads(result["auth_data"])
             result["version"] = str(result["version"])
+            if not result["hash"]:
+                result["hash"] = 0
```
[the remark this answers is not in the document]
Comment (uhoreg) [reply synapse/storage/e2e_room_keys.py]: It's stored as a string because that's what the CS API expects (it's an opaque string), but it's always an int.

**Comment by uhoreg** [comment]
The mypy buildkite failure looks weird, and I'm guessing it's either not-my-fault, or something that will be fixed when I merge in develop (after fixing the merge conflicts).

File: synapse/storage/e2e_room_keys.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (uhoreg) [reply synapse/storage/e2e_room_keys.py]: Changed the DB to store an int, and have the endpoints convert to string.

File: synapse/handlers/e2e_room_keys.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (uhoreg) [reply synapse/handlers/e2e_room_keys.py]: changed to "etag", as that's what the MSC says now

File: synapse/storage/data_stores/main/e2e_room_keys.py [hunk 7]
```diff
@@ -162,6 +152,95 @@ def get_e2e_room_keys(self, user_id, version, room_id=None, session_id=None):
...
 
+    def get_e2e_room_keys_multi(self, user_id, version, room_keys):
+        """Get multiple room keys at a time.  The difference between this function and
+        get_e2e_room_keys is that this function can be used to retrieve
+        multiple specific keys at a time, whereas get_e2e_room_keys is used for
+        getting all the keys in a backup version, all the keys for a room, or a
+        specific key.
+
+        Args:
+            user_id(str): the user whose backup we're querying
```
Comment (richvdh) [inline synapse/storage/data_stores/main/e2e_room_keys.py]: largely for future reference: by convention, there is a space between the param name and the type, though annoyingly the rest of this file doesn't seem to have got the memo :/

File: synapse/storage/data_stores/main/e2e_room_keys.py [hunk 8]
```diff
@@ -162,6 +152,95 @@ def get_e2e_room_keys(self, user_id, version, room_id=None, session_id=None):
...
+            version(str): the version ID of the backup we're querying about
+            room_keys(dict[dict[iterable[str]]]): a map of room IDs to dict which
```
Comment (richvdh) [inline synapse/storage/data_stores/main/e2e_room_keys.py]: the type syntax includes the key type: `dict[str, dict[str, iterable[str]]]`

We can maybe simplify the description:

```
            room_keys (dict[str, dict[str, iterable[str]]]): 
                map from roomID -> {"session": [session ids]}
```

File: synapse/storage/data_stores/main/e2e_room_keys.py [hunk 9]
```diff
@@ -162,6 +152,95 @@ def get_e2e_room_keys(self, user_id, version, room_id=None, session_id=None):
...
+                has a "session" key that is an iterable of session IDs that we
+                want to query
+
+        Returns:
+           Deferred[dict[dict[dict]]]: a map of room IDs to session IDs to room key
```
Comment (richvdh) [inline synapse/storage/data_stores/main/e2e_room_keys.py]:
```suggestion
           Deferred[dict[str, dict[str, dict]]]: a map of room IDs to session IDs to room key
```

File: synapse/storage/data_stores/main/e2e_room_keys.py [hunk 10]
```diff
@@ -236,10 +316,12 @@ def _get_e2e_room_keys_version_info_txn(txn):
                 txn,
                 table="e2e_room_keys_versions",
                 keyvalues={"user_id": user_id, "version": this_version, "deleted": 0},
-                retcols=("version", "algorithm", "auth_data"),
+                retcols=("version", "algorithm", "auth_data", "etag"),
             )
             result["auth_data"] = json.loads(result["auth_data"])
             result["version"] = str(result["version"])
+            if not result["etag"]:
```
Comment (richvdh) [inline synapse/storage/data_stores/main/e2e_room_keys.py]: an explict `if result["etag"] is None` would make the intention here clearer, though in practice it makes little difference.

File: synapse/storage/data_stores/main/e2e_room_keys.py [hunk 11]
```diff
@@ -288,21 +370,31 @@ def _create_e2e_room_keys_version_txn(txn):
...
+    def update_e2e_room_keys_version(
+        self, user_id, version, info=None, version_etag=None
+    ):
         """Update a given backup version
 
         Args:
             user_id(str): the user whose backup version we're updating
             version(str): the version ID of the backup version we're updating
             info(dict): the new backup version info to store
+            version_etag(int): tag of the keys in the backup
```
Comment (richvdh) [inline synapse/storage/data_stores/main/e2e_room_keys.py]:
```suggestion
            version_etag (Optional[int]): tag of the keys in the backup
```

File: synapse/storage/data_stores/main/e2e_room_keys.py [hunk 12]
```diff
@@ -288,21 +370,31 @@ def _create_e2e_room_keys_version_txn(txn):
         )
 
     @trace
-    def update_e2e_room_keys_version(self, user_id, version, info):
+    def update_e2e_room_keys_version(
+        self, user_id, version, info=None, version_etag=None
```
Comment (richvdh) [inline synapse/storage/data_stores/main/e2e_room_keys.py]: if `info` is now allowed to be None, please update the type in docstring accordingly. what happens if it's not set?

File: synapse/storage/data_stores/main/e2e_room_keys.py (hunk 11 above)
Comment (richvdh) [reply synapse/storage/data_stores/main/e2e_room_keys.py]: what does it mean for this to be None?

File: synapse/storage/data_stores/main/e2e_room_keys.py [hunk 13]
```diff
@@ -288,21 +370,31 @@ def _create_e2e_room_keys_version_txn(txn):
...
 
-        return self._simple_update(
-            table="e2e_room_keys_versions",
-            keyvalues={"user_id": user_id, "version": version},
-            updatevalues={"auth_data": json.dumps(info["auth_data"])},
-            desc="update_e2e_room_keys_version",
-        )
+        if info and "auth_data" in info:
+            updatevalues["auth_data"] = json.dumps(info["auth_data"])
+        if version_etag:
```
Comment (richvdh) [inline synapse/storage/data_stores/main/e2e_room_keys.py]: surely it's not correct to handle zero differently to other `int`s?

```suggestion
        if version_etag is not None:
```

File: synapse/storage/data_stores/main/e2e_room_keys.py [hunk 14]
```diff
@@ -78,29 +38,59 @@ def set_e2e_room_key(self, user_id, version, room_id, session_id, room_key):
...
+
+    @defer.inlineCallbacks
+    def add_e2e_room_keys(self, user_id, version, room_keys):
+        """Bulk add room keys to a given backup.
+
+        Args:
+            user_id(str): the user whose backup we're adding to
+            version(str): the version ID of the backup for the set of keys we're adding to
+            room_keys(iterable[(str, str, dict)]): the keys to add, in the form
+                (roomID, sessionID, keyData)
```
Comment (richvdh) [inline synapse/storage/data_stores/main/e2e_room_keys.py]: I don't love the fact that we have a storage-layer method making undocumented assumptions about the shape of the key data dict, but :man_shrugging:
